### ISSUE TITLE
Saving checkpoint has significantly increased after updating to 1.5.2

### ISSUE THREAD

**Issue by ChickenTarm** [PyTorchLightning/pytorch-lightning#10643]
## 🐛 Bug

I recently upgraded from 1.4.9 -> 1.5.2. And I noticed that `save_checkpoint` was taking significantly longer ~ on the scale of hours. Namely the recursive state_dict traversal, `_recurse_update_dict`, is what takes up the chunk of time. I also tested this with 1.5.0. Also I tested this with both torch versions 1.10.0 and 1.9.1.

What I can say that might increase the traversal time is that almost all my modules are nested inside a `torch.nn.modules.container.ModuleDict` and those are also nested inside a `torch.nn.modules.container.ModuleDict` and so on. This is because I have a recursive model builder

### To Reproduce
I waited for 10 mins for the 100 to finish, but i just killed it instead.

`

    import torch
    import time
    from torch import nn
    import torch.nn.functional as F
    import pytorch_lightning as pl
    from torch.nn.modules.container import ModuleDict
    
    
    class LitAutoEncoder(pl.LightningModule):
        def __init__(self, num_dicts):
            super().__init__()
            self.num_dict = num_dicts
            if num_dicts == 0:
                self.encoder = nn.Sequential(nn.Linear(28 * 28, 64), nn.ReLU(), nn.Linear(64, 3))
                self.decoder = nn.Sequential(nn.Linear(3, 64), nn.ReLU(), nn.Linear(64, 28 * 28))
            else:
    
                def recursive_build_encoder(stop_num, current_num):
                    if stop_num == current_num:
                        return nn.Sequential(nn.Linear(28 * 28, 64), nn.ReLU(), nn.Linear(64, 3))
                    else:
                        module_dict = ModuleDict()
                        module_dict[str(current_num)] = recursive_build_encoder(stop_num, current_num + 1)
                        return module_dict
    
                def recursive_build_decoder(stop_num, current_num):
                    if stop_num == current_num:
                        return nn.Sequential(nn.Linear(3, 64), nn.ReLU(), nn.Linear(64, 28 * 28))
                    else:
                        module_dict = ModuleDict()
                        module_dict[str(current_num)] = recursive_build_decoder(stop_num, current_num + 1)
                        return module_dict
    
                self.encoder = recursive_build_encoder(self.num_dict, 0)
                self.decoder = recursive_build_decoder(self.num_dict, 0)
    
        def forward(self, x):
            # in lightning, forward defines the prediction/inference actions
            embedding = self.encoder(x)
            return embedding
    
        def training_step(self, batch, batch_idx):
            # training_step defined the train loop.
            # It is independent of forward
            x, y = batch
            x = x.view(x.size(0), -1)
            z = self.encoder(x)
            x_hat = self.decoder(z)
            loss = F.mse_loss(x_hat, x)
            # Logging to TensorBoard by default
            self.log("train_loss", loss)
            return loss
    
        def configure_optimizers(self):
            optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
            return optimizer
    
    
    start_0_time = time.time()
    LitAutoEncoder(num_dicts=0).state_dict()
    end_0_time = time.time()
    
    print(f"0: {end_0_time - start_0_time}")
    
    start_10_time = time.time()
    LitAutoEncoder(num_dicts=10).state_dict()
    end_10_time = time.time()
    
    print(f"10: {end_10_time - start_10_time}")
    
    start_100_time = time.time()
    LitAutoEncoder(num_dicts=100).state_dict()
    end_100_time = time.time()
    
    print(f"100: {end_100_time - start_100_time}")

`

### Environment

* CUDA:
        - GPU:
                - Tesla T4
        - available:         True
        - version:           10.2
* Packages:
        - numpy:             1.21.4
        - pyTorch_debug:     False
        - pyTorch_version:   1.10.0+cu102
        - pytorch-lightning: 1.5.2
        - tqdm:              4.62.3
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.9.7
        - version:           #61~18.04.1-Ubuntu SMP Thu Oct 28 04:16:28 UTC 2021

cc @awaelchli @ananthsub @ninginthecloud

**Comment by ananthsub** [comment]
@daniellepintz @jjenniferdai it'd be good to add profiling for all the parts of checkpoint saving to identify these bottlenecks more easily

**Comment by awaelchli** [comment]
Thanks for this runnable example, I can reproduce. We don't override the state_dict() methods of nn.Module, but so far I was able to identify the call to

[`self._register_sharded_tensor_state_dict_hooks_if_available()
`](https://github.com/PyTorchLightning/pytorch-lightning/blob/af0bb96f0ff645102680a7adc99dc131cfeb9c0b/pytorch_lightning/core/lightning.py#L121)

in the `__init__` of LightningModule is causing this. 

So the temporary workaround for you to get unblocked is to basically disable this
```python
LitAutoEncoder._register_sharded_tensor_state_dict_hooks_if_available = lambda *_: None
```

**Comment by ananthsub** [comment]
@pritamdamani87 have you seen regressions elsewhere with using sharded tensors like this?

**Comment by ChickenTarm** [comment]
> Thanks for this runnable example, I can reproduce. We don't override the state_dict() methods of nn.Module, but so far I was able to identify the call to
> 
> [`self._register_sharded_tensor_state_dict_hooks_if_available() `](https://github.com/PyTorchLightning/pytorch-lightning/blob/af0bb96f0ff645102680a7adc99dc131cfeb9c0b/pytorch_lightning/core/lightning.py#L121)
> 
> in the `__init__` of LightningModule is causing this.
> 
> So the temporary workaround for you to get unblocked is to basically disable this
> 
> ```python
> LitAutoEncoder._register_sharded_tensor_state_dict_hooks_if_available = lambda *_: None
> ```

What would the implications of this be if I used multi gpu sharded training?

**Comment by ananthsub** [comment]
> What would the implications of this be if I used multi gpu sharded training?

Assuming by `multi gpu sharded training` you mean Fairsale ShardedDataParallel, there should be no impact.

**Comment by awaelchli** [comment]
I believe the implementation of this function is inefficient as it redundantly recurses:

https://github.com/pytorch/pytorch/blob/c41d8290b382f90318cc5f2a90e48eaa37023571/torch/distributed/_sharded_tensor/__init__.py#L375-L382

```python
def _recurse_update_dict(module, destination, prefix):
    for attr_name, attr in module.__dict__.items():
        if isinstance(attr, ShardedTensor):
            destination[prefix + attr_name] = attr

    for submodule_name, submodule in module.named_modules():
        if submodule_name != '':
            _recurse_update_dict(submodule, destination, prefix + submodule_name + '.')
```
From what I can tell, the recursion is not necessary, because `module.named_modules()` ALREADY returns all the nested submodules in a generator. I would instead propose this: 

```python
def _recurse_update_dict(module, destination, prefix):
    for submodule_name, submodule in module.named_modules():
        for attr_name, attr in module.__dict__.items():
            if isinstance(attr, ShardedTensor):
                destination[prefix + submodule_name + "." + attr_name] = attr
```

I have opened an issue on the PyTorch side https://github.com/pytorch/pytorch/issues/68805 and will probably submit a PR.

**Comment by ChickenTarm** [comment]
> 

I tried this workaround, and I noticed that the time it takes to save still takes a long time.

**Comment by awaelchli** [comment]
Not with the code that you provided. I added the line here: 

```python
import torch
import time
from torch import nn
import torch.nn.functional as F
import pytorch_lightning as pl
from torch.nn.modules.container import ModuleDict


class LitAutoEncoder(pl.LightningModule):
    def __init__(self, num_dicts):
        super().__init__()
        self.num_dict = num_dicts
        if num_dicts == 0:
            self.encoder = nn.Sequential(nn.Linear(28 * 28, 64), nn.ReLU(), nn.Linear(64, 3))
            self.decoder = nn.Sequential(nn.Linear(3, 64), nn.ReLU(), nn.Linear(64, 28 * 28))
        else:

            def recursive_build_encoder(stop_num, current_num):
                if stop_num == current_num:
                    return nn.Sequential(nn.Linear(28 * 28, 64), nn.ReLU(), nn.Linear(64, 3))
                else:
                    module_dict = ModuleDict()
                    module_dict[str(current_num)] = recursive_build_encoder(stop_num, current_num + 1)
                    return module_dict

            def recursive_build_decoder(stop_num, current_num):
                if stop_num == current_num:
                    return nn.Sequential(nn.Linear(3, 64), nn.ReLU(), nn.Linear(64, 28 * 28))
                else:
                    module_dict = ModuleDict()
                    module_dict[str(current_num)] = recursive_build_decoder(stop_num, current_num + 1)
                    return module_dict

            self.encoder = recursive_build_encoder(self.num_dict, 0)
            self.decoder = recursive_build_decoder(self.num_dict, 0)

    def forward(self, x):
        # in lightning, forward defines the prediction/inference actions
        embedding = self.encoder(x)
        return embedding

    def training_step(self, batch, batch_idx):
        # training_step defined the train loop.
        # It is independent of forward
        x, y = batch
        x = x.view(x.size(0), -1)
        z = self.encoder(x)
        x_hat = self.decoder(z)
        loss = F.mse_loss(x_hat, x)
        # Logging to TensorBoard by default
        self.log("train_loss", loss)
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        return optimizer

# ADDED HERE
LitAutoEncoder._register_sharded_tensor_state_dict_hooks_if_available = lambda *_: None


start_0_time = time.time()
LitAutoEncoder(num_dicts=0).state_dict()
end_0_time = time.time()

print(f"0: {end_0_time - start_0_time}")

start_10_time = time.time()
LitAutoEncoder(num_dicts=10).state_dict()
end_10_time = time.time()

print(f"10: {end_10_time - start_10_time}")

start_100_time = time.time()
LitAutoEncoder(num_dicts=100).state_dict()
end_100_time = time.time()

print(f"100: {end_100_time - start_100_time}")

```

This retrieves the state dict instantly (PL 1.5.3, torch 1.10)

**Comment by ChickenTarm** [comment]
So you are correct that the state_dict retrieval is fast, but the checkpointing still appears to be slow. Here is another working example of this slowdown.

`

    import os
    
    import torch
    from torch.utils.data import DataLoader, Dataset
    from pytorch_lightning.callbacks.model_checkpoint import ModelCheckpoint
    import time
    from torch.nn.modules.container import ModuleDict
    
    from pytorch_lightning import LightningModule, Trainer
    
    
    class RandomDataset(Dataset):
        def __init__(self, size, length):
            self.len = length
            self.data = torch.randn(length, size)
    
        def __getitem__(self, index):
            return self.data[index]
    
        def __len__(self):
            return self.len
    
    
    def recursive_forward(x, layer: ModuleDict, current_layer, stop_layer):
        if current_layer == stop_layer:
            return layer(x)
        else:
            return recursive_forward(x, layer[str(current_layer)], current_layer + 1, stop_layer)
    
    
    class BoringModel(LightningModule):
        def __init__(self, num_dicts: int):
            super().__init__()
            self.num_dicts = num_dicts
            if num_dicts == 0:
                self.layer = torch.nn.Linear(1024, 2)
            else:
    
                def recursive_build_layer(stop_num, current_num):
                    if stop_num == current_num:
                        return torch.nn.Linear(1024, 2)
                    else:
                        module_dict = ModuleDict()
                        module_dict[str(current_num)] = recursive_build_layer(stop_num, current_num + 1)
                        return module_dict
    
                self.layer = recursive_build_layer(self.num_dicts, 0)
    
        def forward(self, x):
            if self.num_dicts > 0:
                return recursive_forward(x, self.layer, 0, self.num_dicts)
            else:
                return self.layer(x)
    
        def training_step(self, batch, batch_idx):
            loss = self(batch).sum()
            self.log("train_loss", loss)
            return {"loss": loss}
    
        def validation_step(self, batch, batch_idx):
            loss = self(batch).sum()
            self.log("valid_loss", loss)
    
        def test_step(self, batch, batch_idx):
            loss = self(batch).sum()
            self.log("test_loss", loss)
    
        def configure_optimizers(self):
            return torch.optim.SGD(self.layer.parameters(), lr=0.1)
    
    
    def iterate_over_dicts(num_dicts: int):
        train_data = DataLoader(RandomDataset(1024, 64), batch_size=2)
        val_data = DataLoader(RandomDataset(1024, 64), batch_size=2)
        test_data = DataLoader(RandomDataset(1024, 64), batch_size=2)
    
        model = BoringModel(num_dicts=num_dicts)
        model._register_sharded_tensor_state_dict_hooks_if_available = lambda *_: None
        trainer = Trainer(
            default_root_dir=os.getcwd(),
            limit_train_batches=1,
            limit_val_batches=1,
            limit_test_batches=1,
            num_sanity_val_steps=0,
            max_epochs=1,
            enable_model_summary=False,
        )
        trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    
        checkpointer = ModelCheckpoint()
    
        start_time = time.time()
        checkpointer.save_checkpoint(trainer)
        end_time = time.time()
    
        print(f"{num_dicts}: {end_time - start_time}")
    
    
    def run():
        num_dicts_list = [0, 10, 20, 100]
        for num_dicts in num_dicts_list:
            iterate_over_dicts(num_dicts=num_dicts)
    
    
    if __name__ == "__main__":
        run()


`

**Comment by awaelchli** [comment]
Nono, you have to do it like I showed:
```py
BoringModel._register_sharded_tensor_state_dict_hooks_if_available = lambda *_: None
model = BoringModel(num_dicts=num_dicts)
```
NOT this
```py
model = BoringModel(num_dicts=num_dicts)
model._register_sharded_tensor_state_dict_hooks_if_available = lambda *_: None
```

Otherwise the patch will have no effect. Sorry for the delayed answer. 
PyTorch hasn't prioritized my fix. I hope to get it in soon.

**Comment by awaelchli** [comment]
Fix in pytorch has landed: https://github.com/pytorch/pytorch/commit/603a1de871ddaae39ba746ac652da23a9f568d40
Shortly you will be able to use the pytorch nightly build with this fix if you wish to, or continue to  use the workaround above until pytorch 1.11 is released. 

Thanks for reporting this @ChickenTarm
